lua/arc.lua

```diff
@@ -0,0 +1,244 @@
+--- Arc class
+-- @module arc
+-- @alias Arc
+require 'norns'
+
+
+---------------------------------
+-- Arc device class
+
+local Arc = {}
+Arc.devices = {}
+Arc.list = {}
+Arc.vport = {}
+for i=1,4 do
+  Arc.vport[i] = {
+    name = "none",
+    delta_callbacks = {},
+    key_callbacks = {},
+    index = 0,
+    led = function() end,
+    all = function() end,
+    refresh = function() end,
+    attached = false
+  }
+end
+Arc.__index = Arc
+
+--- constructor
+-- @tparam integer id : arbitrary numeric identifier
+-- @tparam string serial : serial
+-- @tparam string name : name
+-- @tparam userdata dev : opaque pointer to device
+function Arc.new(id, serial, name, dev)
+  local device = setmetatable({}, Arc)
+  device.id = id
+  device.serial = serial
+  name = name .. " " .. serial
+  --while tab.contains(Arc.list, name) do
+  --  name = name .. "+"
+  --end
+  device.name = name
+  device.dev = dev -- opaque pointer
+  device.delta = nil -- delta event callback
+  device.key = nil -- key event callback
+  device.remove = nil -- device unplug callback
+  device.ports = {} -- list of virtual ports this device is attached to
+
+  -- autofill next postiion
+  local connected = {}
+  for i=1,4 do
+    table.insert(connected, Arc.vport[i].name)
+  end
+  if not tab.contains(connected, name) then
+    for i=1,4 do
+      if Arc.vport[i].name == "none" then
+        Arc.vport[i].name = name
+        break
+      end
+    end
+  end
+
+  return device
+end
+
+--- static callback when any arc device is added;
+-- user scripts can redefine
+-- @param dev : a Arc table
+function Arc.add(dev)
+  print("arc added:", dev.id, dev.name, dev.serial)
+end
+
+--- scan device list and grab one, redefined later
+function Arc.reconnect() end
+
+--- static callback when any arc device is removed;
+-- user scripts can redefine
+-- @param dev : a Arc table
+function Arc.remove(dev) end
+
+--- set state of single LED on this arc device
+-- @tparam integer ring : ring index (1-based!)
+-- @tparam integer x : led index (1-based!)
+-- @tparam integer val : LED brightness in [0, 15]
+function Arc:led(ring, x, val)
+  arc_set_led(self.dev, ring, x, val)
+end
+
+--- set state of all LEDs on this arc device
+-- @tparam integer val : LED brightness in [0, 15]
+function Arc:all(val)
+  arc_all_led(self.dev, val)
+end
+
+--- update any dirty quads on this arc device
+function Arc:refresh()
+  monome_refresh(self.dev)
+end
+
+--- print a description of this arc device
+function Arc:print()
+  for k, v in pairs(self) do
+    print('>> ', k, v)
+  end
+end
+
+--- create device, returns object with handler and send
+function Arc.connect(n)
+  local n = n or 1
+  if n > 4 then n = 4 end
+
+  Arc.vport[n].index = Arc.vport[n].index + 1
+
+  local d = {
+    index = Arc.vport[n].index,
+    port = n,
+    delta = function(n, delta)
+      print("arc input")
```

Comment:
should this (and line 120) get commented out for release?

Reply:
yeah, but it's identical to the grid code in the next module, so i'm fine with it.